### PR TITLE
Adds cert-manager add-on to Keip operator

### PR DESCRIPTION
**Pull request by figliold**
Adds cert-manager add-on to Keip operator.

Provides an add-on that creates a [cert-manager.io/v1.Certificate](https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1.Certificate) based on annotations in an `IntegrationRoute`.

## Testing

Below are a few example tests to test integration routes with the certmanager addon installed and the certmanager addon uninstalled.

### Basic Integration Route (no certmanager addon installed)

**NOTE:**  Set up Python virtual environment (See `operator/webhook/README.md` for directions).

```
operator > cd webhook
webhook > docker build --tag ghcr.io/octoconsulting/keip/route-webhook:0.11.0 .
webhook > cd ..
operator > make metacontroller/deploy
operator > make controller/deploy
operator > cd example
example > kubectl apply -k integration-routes/basic
```

Follow the directions in section `Example Basic Integration Route` in `operator/example/README.md` for testing directions.

#### Clean up
```
example > kubectl delete -k integration-routes/basic --ignore-not-found
example > cd ..
operator > make controller/undeploy
operator > make metacontroller/undeploy
```

### Basic Integration Route (certmanager addon installed)

```
operator > cd webhook
webhook > docker build --tag ghcr.io/octoconsulting/keip/route-webhook:0.11.0 .
webhook > cd ..
operator > make metacontroller/deploy
operator > make controller/deploy
operator > make addons/certmanager/deploy
operator > cd example
example > kubectl apply -k bases/certmanager
example > kubectl apply -k bases/certmanager/issuer
example > kubectl apply -k integration-routes/basic
```

**NOTE:** If you see the following error when running the command `kubectl apply -k example/cert-manager/issuer`, wait a bit and try again.
```
 tls: failed to verify certificate: x509: certificate signed by unknown authority
```

Follow the directions in section `Example Basic Integration Route` in `operator/example/README.md` for testing directions.

#### Clean up
```
example > kubectl delete -k integration-routes/basic
example > kubectl delete -k bases/certmanager/issuer
example > kubectl delete -k bases/certmanager
example > cd ..
operator > make addons/certmanager/undeploy
operator > make controller/undeploy
operator > make metacontroller/undeploy
```

### TLS Integration Route (certmanager addon installed)

```
operator > cd webhook
webhook > docker build --tag ghcr.io/octoconsulting/keip/route-webhook:0.11.0 .
webhook > cd ..
operator > make metacontroller/deploy
operator > make controller/deploy
operator > make addons/certmanager/deploy
operator > cd example
example > kubectl apply -k bases/certmanager
example > kubectl apply -k bases/certmanager/issuer
example > kubectl apply -k integration-routes/certmanager-addon
```

**NOTE:** If you see the following error when running the command `kubectl apply -k example/cert-manager/issuer`, wait a bit and try again.
```
 tls: failed to verify certificate: x509: certificate signed by unknown authority
```

Follow the directions in section `Example IntegrationRoute Using TLS and the CertManager Addon` in `operator/example/README.md` for testing directions.

#### Clean up
```
example > kubectl delete -k integration-routes/certmanager-addon
example > kubectl delete secret -n default testroute-certstore
example > kubectl delete -k bases/certmanager/issuer
example > kubectl delete -k bases/certmanager
example > cd ..
operator > make addons/certmanager/undeploy
operator > make controller/undeploy
operator > make metacontroller/undeploy
```

## Run the unit tests

**NOTE:** Must have a Python virtual environment to run the unit tests (See `operator/webhook/README.md` for directions).

### Windows
```
(keip) webhook > make win-test
```
### Mac/Unix/Linux
```
(keip) webhook > make test
```

### CODE REVIEW

File: operator/webhook/addons/certmanager/main.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            l.strip()
+            for l in annotations.get("cert-manager.io/subject-localities").split(",")
+        ]
+        if annotations.get("cert-manager.io/subject-localities", None) is not None
+        else []
+    )
+
+    dns_names = [
+        f"{annotations['cert-manager.io/common-name']}.{metadata['namespace']}.svc.cluster.local",
+        f"{annotations['cert-manager.io/common-name']}.{metadata['namespace']}.svc",
```
Comment (jhunzik) [inline operator/webhook/addons/certmanager/main.py:82]: Is it worth adding this to be thorough?
```suggestion
        f"{annotations['cert-manager.io/common-name']}.{metadata['namespace']}",

```

File: operator/webhook/addons/certmanager/main.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        [
+            l.strip()
+            for l in annotations.get("cert-manager.io/subject-localities").split(",")
+        ]
+        if annotations.get("cert-manager.io/subject-localities", None) is not None
+        else []
+    )
+
+    dns_names = [
+        f"{annotations['cert-manager.io/common-name']}.{metadata['namespace']}.svc.cluster.local",
```
Comment (jhunzik) [inline operator/webhook/addons/certmanager/main.py:81]: :pencil2: You could use `{common-name}` for these since you already extract it above.

**Review by jhunzik — APPROVED** [review]

File: operator/webhook/addons/certmanager/main.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+def _new_certificate(object) -> Mapping:
+    metadata = object.get("metadata", None)
+    if metadata is None:
+        _LOGGER.error("IntegrationRoute does not contain metadata.")
+        return {}
+
+    annotations = metadata.get("annotations", None)
+    if annotations is None:
+        _LOGGER.error("IntegrationRoute does not contain metadata.annotations.")
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/main.py:19]: :question: Should this be logged as an error? It should be acceptable for an integration-route not to require a cert. How about a `debug` log that states there is nothing to be done instead?

File: operator/webhook/app.py [hunk 4]
```diff
@@ -9,37 +11,85 @@
...
+COMPOSITE_CONTROLLER = "CompositeController"
+
+DECORATOR_CONTROLLER = "DecoratorController"
+
+INTEGRATION_ROUTES_RESOURCE = "integrationroutes"
+
+CERTIFICATES_RESOURCE = "certificates"
+
 
 async def webhook(request: Request):
```
Comment (a-asaad) [inline operator/webhook/app.py:29]: I think we might be over-complicating here. As far as this method is concerned, it should really just be dealing with  serializing/deserializing the JSON, and providing a catch-all for errors that bubble up.

We could simplify by treating the request body as opaque and passing that to the appropriate sync handlers. Each handler should already know the type of request it receives (Composite vs Decorator), so we can let them pick out the specific fields, rather than do it here. Plus they might need other fields in the future, so it makes sense to pass along the full request.

File: operator/Makefile [hunk 5]
```diff
@@ -18,8 +18,14 @@ get-tag:
...
 	rm -rf output
-	mkdir output
-	kustomize build ./controller > ./output/controller.yaml
+	mkdir -p output/addons
+	echo "Running kustomize build on ./controller"
+	kustomize build ./controller > ./output/core-controller.yaml
+	for dir in ./controller/addons/*/ ; do \
+		echo "Running kustomize build on $$dir" ; \
+		echo "---" >> ./output/addons/addons-controller.yaml ; \
+		kustomize build $$dir >> ./output/addons/addons-controller.yaml ; \
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/Makefile]: Changed

File: operator/Makefile [hunk 6]
```diff
@@ -1,6 +1,6 @@
-VERSION ?= 0.6.0
+VERSION ?= 0.7.0
 GIT_TAG := operator_v$(VERSION)
-KEIP_INTEGRATION_IMAGE ?= ghcr.io/octoconsulting/keip/minimal-app:latest
+KEIP_INTEGRATION_IMAGE ?= ghcr.io/octoconsulting/keip/minimal-app:0.0.4
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/Makefile:4]: Added TODO

File: operator/Makefile [hunk 7]
```diff
@@ -34,8 +40,14 @@ controller/deploy: crd/deploy
 	$(KUBECTL) -n $(CONTROLLER_NAMESPACE) create cm keip-controller-props --from-literal=integration-image=$(KEIP_INTEGRATION_IMAGE) -o yaml --dry-run=client | $(KUBECTL) apply -f -
 
 controller/undeploy: crd/undeploy
-	-$(KUBECTL_DELETE) -n $(CONTROLLER_NAMESPACE) cm keip-controller-props
-	-$(KUBECTL_DELETE) -k controller
+	$(KUBECTL_DELETE) -n $(CONTROLLER_NAMESPACE) cm keip-controller-props
+	$(KUBECTL_DELETE) -k controller
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/Makefile]: Good catch. Fixed.

File: operator/controller/addons/certmanager/certmanager-controller.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  name: certmanager-controller
+spec:
+  resources:
+    - apiVersion: keip.octo.com/v1alpha1
+      resource: integrationroutes
+  attachments:
+    - apiVersion: cert-manager.io/v1
+      resource: certificates
+      updateStrategy:
+        method: RollingRecreate
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/controller/addons/certmanager/certmanager-controller.yaml]: Changed

File: operator/controller/core-privileges.yaml [hunk 9]
```diff
@@ -9,7 +9,7 @@ metadata:
   name: spring-cloud-kubernetes
 rules:
   - apiGroups: [""]
-    resources: ["configmaps"]
+    resources: ["configmaps", "pods"]
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/controller/core-privileges.yaml:12]: Yes, I saw an error in the pod logs when testing the actuators.

```
403 forbidden when testing /actuator/health Failed to get pod with name:. 
You should look into this if things aren't working as you expect. 
Are you missing service account permissions?
```

File: operator/webhook/addons/certmanager/main.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+def _get_keystore_type(object) -> str:
+    return object["spec"]["tls"]["keystore"]["type"]
+
+
+def _get_password_secret_ref_name(object) -> str:
+    return object["spec"]["tls"]["keystore"]["passwordSecretRef"]
+
+
+def sync_certificate(object) -> Mapping:
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Change to `obj`.

File: operator/webhook/addons/certmanager/main.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+import os
+from typing import Mapping, List
+
+_LOGGER = logging.getLogger(__name__)
+
+from webhook.core.sync import sync
+
+
+def _new_certificate(object) -> Mapping:
+    metadata = object.get("metadata", None)
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Changed

File: operator/webhook/addons/certmanager/main.py (hunk 3 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Changed

File: operator/webhook/app.py (hunk 4 above)
Comment (figliold) [reply operator/webhook/app.py]: Fixed

File: operator/example/README.md [hunk 12]
```diff
@@ -5,54 +5,109 @@ Deploys a pod that runs a simple Spring Integration app that periodically logs a
 Prerequisites:
 
 - Kubernetes cluster (v1.24+)
-- Metacontroller
-- Keip CRDs and controller
+- Metacontroller (see `metacontroller/deploy` in `operator/Makefile`)
+- Keip CRDs and controller (see `controller/deploy` in `operator/Makefile`)
 - Access to a `keip-integration` image
+- Certmanager addon (see `addons/certmanager/deploy` in `operator/Makefile`)
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/example/README.md]: Added

File: operator/webhook/Makefile [hunk 13]
```diff
@@ -28,27 +28,27 @@ start-dev-server:
 
 .PHONY: test
 test:
-	cd test && mkdir -p $(TEST_COVERAGE_DIR) && coverage run --data-file=$(TEST_COVERAGE_FILE) -m pytest -vv
+	rm -rf $(TEST_COVERAGE_DIR) && mkdir -p $(TEST_COVERAGE_DIR) && coverage run --data-file=$(TEST_COVERAGE_FILE) -m pytest -vv
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/Makefile]: Fixed

File: operator/webhook/addons/certmanager/main.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
+import logging
+import sys
+import os
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Removed

File: operator/webhook/addons/certmanager/main.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
+import logging
+import sys
+import os
+from typing import Mapping, List
+
+_LOGGER = logging.getLogger(__name__)
+
+from webhook.core.sync import sync
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Removed

File: operator/webhook/addons/certmanager/main.py [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+                "IntegrationRoute does not contain metadata.annotations.cert-manager.io/cluster-issuer"
+            )
+            return {}
+
+        common_name = annotations.get("cert-manager.io/common-name", None)
+        if common_name is None:
+            _LOGGER.error(
+                "IntegrationRoute does not contain metadata.annotations.cert-manager.io/common-name"
+            )
+            return {}
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/main.py:38]: :question: Instead of treating this as an error, does it make sense to use the integration-route's name as a default for `common_name`?

File: operator/webhook/addons/certmanager/main.py [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+        return {}
+
+    keystore_type = _get_keystore_type(obj)
+    password_secret_ref_name = _get_password_secret_ref_name(obj)
+
+    alt_names = (
+        [n.strip() for n in annotations.get("cert-manager.io/alt-names").split(",")]
+        if annotations.get("cert-manager.io/alt-names", None) is not None
+        else []
+    )
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/main.py:49]: So for this annotation: `"cert-manager.io/alt-names": ","`

Seems this logic would end up adding two empty strings to the list of `dnsNames`.  Is that the expected behavior?

An alternative for filtering blank strings could be:
```
alt_names = [n.strip() for n in annotations.get("cert-manager.io/alt-names", []).split(",") if n.strip()]
```

If calling `strip()` twice bothers you, you could get fancier:
```
alt_names = [stripped_name for n in annotations.get("cert-manager.io/alt-names", []).split(",") if (stripped_name := n.strip())]
```

Comment (a-asaad) [reply operator/webhook/addons/certmanager/main.py:49]: As this string splitting logic is repeated across multiple annotations, might make sense to extract it to a common function, and possibly add some unit tests for some of the obvious edge cases (no need to go overboard).

File: operator/webhook/core/sync.py [hunk 18]
```diff
@@ -426,44 +411,49 @@ def _new_actuator_service(parent):
...
 
+
 def _gen_children(parent) -> List[Mapping]:
     return [_new_deployment(parent), _new_actuator_service(parent)]
 
 
-def sync(parent) -> Mapping:
+def sync(body) -> Mapping:
+    # Request API at https://metacontroller.github.io/metacontroller/api/compositecontroller.html#sync-hook-request
+    _LOGGER.debug("\n\nRequest:\n%s", body)
```
Comment (a-asaad) [inline operator/webhook/core/sync.py:454]: Looks like we already log the request as soon as it's received in `app.py`. We might want to avoid logging again here so we don't spam the logs with redundant info.

File: operator/webhook/addons/certmanager/main.py [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+    return obj["spec"]["tls"]["keystore"]["type"]
+
+
+def _get_password_secret_ref_name(obj) -> str:
+    return obj["spec"]["tls"]["keystore"]["passwordSecretRef"]
+
+
+def sync_certificate(body) -> Mapping:
+    # Request API at for DecoratorController at https://metacontroller.github.io/metacontroller/api/decoratorcontroller.html#sync-hook-request
+    _LOGGER.debug("\n\nRequest:\n%s", body)
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/main.py:150]: Same issue with redundant logging here

File: operator/webhook/core/test/test_sync.py [hunk 20]
```diff
@@ -324,17 +326,21 @@ def test_pod_resources_requests_only(full_route):
...
 
-def load_json(filepath: str) -> Mapping:
-    with open(filepath, "r") as f:
-        return json.load(f)
+def load_json_as_dict(filepath: str) -> Mapping:
+    try:
+        with open(filepath, "r") as f:
+            return json.load(f)
+    except FileNotFoundError as e:
+        _LOGGER.error(f"File not found: {filepath}\n{e}")
```
Comment (a-asaad) [inline operator/webhook/core/test/test_sync.py:343]: Is `_LOGGER` defined in this file? Since we control the filepaths for the unit tests, I'd rather we fail fast vs. log an error and continue. Can we drop the `try except` block?

Comment (a-asaad) [reply operator/webhook/core/test/test_sync.py:343]: This block is repeated in the other tests as well, so I would either extract those into a single shared method, or make the same changes there.

**Comment by a-asaad** [comment]
Can we add a small README to `operator/webhook/addons/certmanager`? Just a brief description of the add-on, maybe a list of the supported annotations, and a link to the example.

File: operator/controller/addons/certmanager/certmanager-controller.yaml [hunk 21]
```diff
@@ -0,0 +1,18 @@
+---
+apiVersion: metacontroller.k8s.io/v1alpha1
+kind: DecoratorController
+metadata:
+  name: certmanager-controller
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/controller/addons/certmanager/certmanager-controller.yaml]: Renamed

File: operator/webhook/addons/certmanager/main.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
+import logging
+from typing import Mapping, List
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py:2]: It is being used now with the new logic I've added.

File: operator/webhook/addons/certmanager/main.py (hunk 16 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Yes, Changed

File: operator/webhook/addons/certmanager/main.py (hunk 17 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py:39]: Fixed, moved to a common function, and added tests

File: operator/webhook/core/sync.py (hunk 18 above)
Comment (figliold) [reply operator/webhook/core/sync.py]: Fixed

File: operator/webhook/addons/certmanager/main.py (hunk 19 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Fixed

File: operator/webhook/core/test/test_sync.py (hunk 20 above)
Comment (figliold) [reply operator/webhook/core/test/test_sync.py]: Changed

**Comment by figliold** [comment]
> Can we add a small README to `operator/webhook/addons/certmanager`? Just a brief description of the add-on, maybe a list of the supported annotations, and a link to the example.

I added a README for the Certmanager Addon

File: .github/workflows/minimal-app.yml [hunk 23]
```diff
@@ -95,6 +96,9 @@ jobs:
       - run: docker push ${{ steps.naming-selector.outputs.FULL_IMAGE_NAME }}
```
Comment (pelletier2017) [inline .github/workflows/minimal-app.yml:96]: It looks like you tag latest but don't push latest. This step could be multiple parts in the same step similar to line 59 where it sets the variables since it all essentially does 1 thing which is pushing images. 

```
- run: |
    docker tag ${{ steps.naming-selector.outputs.FULL_IMAGE_NAME }} ${{ steps.naming-selector.outputs.LATEST_FULL_IMAGE_NAME }}
    docker push ${{ steps.naming-selector.outputs.FULL_IMAGE_NAME }}
    docker push ${{ steps.naming-selector.outputs.LATEST_FULL_IMAGE_NAME }}
  if: env.needs_release == 'true'
```

Comment (figliold) [reply .github/workflows/minimal-app.yml]: Changed

File: .github/workflows/operator.yml [hunk 24]
```diff
@@ -83,4 +83,4 @@ jobs:
           name: Release ${{ steps.tag_version.outputs.new_tag }}
           body: ${{ steps.tag_version.outputs.changelog }}
           artifactErrorsFailBuild: true
-          artifacts: ${{ env.WORKING_DIR }}/output/*
+          artifacts: ${{ env.WORKING_DIR }}/output/*.yaml,${{ env.WORKING_DIR }}/output/addons/*.yaml,${{ env.WORKING_DIR }}/output/addons/*/*.yaml
```
[the remark this answers is not in the document]
Comment (figliold) [reply .github/workflows/operator.yml]: Changed

File: operator/webhook/core/sync.py (hunk 18 above)
Comment (a-asaad) [reply operator/webhook/core/sync.py]: Should we remove the other logging statement as well, that logs the reply?

File: operator/webhook/addons/certmanager/main.py (hunk 19 above)
Comment (a-asaad) [reply operator/webhook/addons/certmanager/main.py]: Should we remove the other logging statement as well, that logs the reply?

File: operator/webhook/addons/certmanager/README.md [hunk 25]
```diff
@@ -0,0 +1,15 @@
+# Certmanager Addon
+
+The Certmanager Addon creates certificates for `IntegrationRoute`s based off of annotations in the `IntegrationRoute`.
+
+See the example `IntegrationRoute` in the [README](../../../example/README.md#example-integrationroute-using-tls-and-the-certmanager-addon).
+
+## Supported Annotations
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/README.md:7]: We should consider supporting the `cert-manager.io/issuer` annotation as well, since we might not always deal with a cluster-level issuer.  However, this PR has already grown considerably, so I would leave it as a follow-on.

**Review by a-asaad — APPROVED** [review]

File: operator/Makefile [hunk 26]
```diff
@@ -17,11 +18,22 @@ get-tag:
...
-	ls -al ./output
+	@rm -rf output
+	@mkdir -p output/addons
+	@echo "Running kustomize build on ./controller/"
+	kustomize build ./controller/ > ./output/core-controller.yaml
+	@for dir in ./controller/addons/*/ ; do \
+		echo "Running kustomize build on $$dir" ; \
+		addon=$$(basename $$dir) ; \
+		mkdir -p output/addons/$$addon ; \
+		kustomize build $$dir >> ./output/addons/$$addon/$${addon}-addon-controller.yaml ; \
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/Makefile]: Removed the redundant "addon" from the name.

File: operator/webhook/addons/certmanager/README.md [hunk 27]
```diff
@@ -0,0 +1,15 @@
+# Certmanager Addon
+
+The Certmanager Addon creates certificates for `IntegrationRoute`s based off of annotations in the `IntegrationRoute`.
+
+See the example `IntegrationRoute` in the [README](../../../example/README.md#example-integrationroute-using-tls-and-the-certmanager-addon).
+
+## Supported Annotations
+- **cert-manager.io/cluster-issuer:** The name of a ClusterIssuer to acquire the Certificate required for this Gateway. It does not matter which namespace your Gateway resides, as ClusterIssuers are non-namespaced resources.
```
[the remark this answers is not in the document]
Comment (figliold) [reply operator/webhook/addons/certmanager/README.md]: Changed

File: operator/webhook/addons/certmanager/README.md (hunk 25 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/README.md:7]: Added

File: operator/webhook/core/sync.py (hunk 18 above)
Comment (figliold) [reply operator/webhook/core/sync.py]: Fixed

File: operator/webhook/addons/certmanager/main.py (hunk 19 above)
Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Fixed

File: operator/webhook/addons/certmanager/main.py [hunk 28]
```diff
@@ -0,0 +1,146 @@
...
+def _get_password_secret_ref_name(obj) -> str:
+    return obj["spec"]["tls"]["keystore"]["passwordSecretRef"]
+
+
+def sync_certificate(body) -> Mapping:
+    # Request API at for DecoratorController at https://metacontroller.github.io/metacontroller/api/decoratorcontroller.html#sync-hook-request
+    obj = body["object"]
+    certificate = _new_certificate(obj)
+    attachments = [certificate] if certificate else []
+    desired_state = {"status": {}, "attachments": attachments}
```
Comment (a-asaad) [inline operator/webhook/addons/certmanager/main.py:145]: :exclamation: It's not entirely clear, but from the [Sync Hook Response docs](https://metacontroller.github.io/metacontroller/api/decoratorcontroller.html#sync-hook-response), it seems that setting the `status` field to an empty JSON could potentially clear out the `status` field on the IntegrationRoute, might be best to play it safe and exclude it from the response. We should only set the `status` in the core sync function.

Comment (figliold) [reply operator/webhook/addons/certmanager/main.py]: Removed status from the certmanager addon response
